stravalib/model.py

```diff
@@ -373,32 +581,66 @@ class Athlete(
     membership: Optional[str] = None
     admin: Optional[bool] = None
     owner: Optional[bool] = None
-    subscription_permissions: Optional[list] = None
+    # TODO: can permission take other inputs other than bool?
+    subscription_permissions: Optional[List[bool]] = None
 
     @validator("athlete_type")
-    def to_str_representation(cls, raw_type):
-        # Replaces legacy "ChoicesAttribute" class
+    def to_str_representation(cls, raw_type: int) -> Optional[str]:
```

Comment:
Here the return type is always string, as you wrote in the comment below. The optional is not needed

Reply:
Ok so here i was responding to this error:
```
stravalib/model.py:603: error: Incompatible return value type (got "Optional[str]", expected "str")  [return-value]
```

this error confuses me as i'm not sure how mypy is looking at this method and determine potential outputs. 

when running this IF for some reason raw_type is not 0 or 1 it will return None. i tried this

```
        athlete_type = {0: "cyclist", 1: "runner"}
        assert raw_type in athlete_type.keys()
        return athlete_type.get(raw_type)
```

 as a check but that doesn't seem to make mypy happy. any guidance is appreciated of if it's easier feel free to suggest a change directly / push here. i struggled with this one.

Reply:
Okay, on closer inspection I think there might actually be a small bug in the code/typing. The validator is taking any value and transforming it into a `Optional[str]` but the `athlete_type` attribute on the class is of type `Optional[int]` which does not quite make sense as the value that ends up in the attribute is a string.

I think we need to alter the validator to a pre-validator that converts the int into a string before pydantic runs it's own validation. This way, the athlete type can be the correct type.

My proposed change is below and passes both type checking and unit test. @jsamoocha I'd like to get your opinion on this since this is undocumented behaviour I'm not 100% sure I understand.

```diff
diff --git a/stravalib/model.py b/stravalib/model.py
index cc873f4..335313d 100644
--- a/stravalib/model.py
+++ b/stravalib/model.py
@@ -544,7 +549,7 @@ class Athlete(
 
     # Undocumented attributes:
     # Undocumented attributes:
     is_authenticated: Optional[bool] = None
-    athlete_type: Optional[int] = None
+    athlete_type: Optional[Literal["cyclist", "runner"]] = None
     friend: Optional[str] = None
     follower: Optional[str] = None
     approve_followers: Optional[bool] = None
@@ -581,8 +586,10 @@ class Athlete(
     owner: Optional[bool] = None
     subscription_permissions: Optional[List[bool]] = None
 
-    @validator("athlete_type")
-    def to_str_representation(cls, raw_type: int) -> str:
+    @validator("athlete_type", pre=True)
+    def to_str_representation(
+        cls, raw_type: Any
+    ) -> Optional[Literal["cyclist", "runner"]]:
         """Replaces legacy 'ChoicesAttribute' class.
 
         Parameters
@@ -595,9 +602,16 @@ class Athlete(
         str
             The string representation of the athlete type.
         """
-
-        athlete_type = {0: "cyclist", 1: "runner"}
-        return athlete_type.get(raw_type)
+        if raw_type == 0:
+            return "cyclist"
+        elif raw_type == 1:
+            return "runner"
+        elif raw_type is None:
+            return None
+        else:
+            raise ValueError(
+                "value is not a valid representation ofr athlete type"
+            )
 
     @lazy_property
     def authenticated_athlete(self) -> Athlete:

Reply:
thanks!! If we moved forward with this solution,  the conditional could be a bit simpler if we did this:

```python
# Define dict - this would allow us to add additional types easily if they arise in the future
athlete_type = {0: "cyclist", 1: "runner"}
if raw_type in [0, 1]:
  return athlete_type.get(raw_type)
elif raw_type is None:
   return None
else:
   raise ValueError("raw_value is not a valid representation of athlete type")
```

But also i have a question about how this all works. When we run mypy is pydantic also running type checks?

Reply:
it looks like we only need to return none - UNLESS there is some instance where a value error would get raised that i'm missing?

Reply:
I agree the correct (user-facing) type of that field should be `Optional[Literal["cyclist", "runner"]]`.

As this is an undocumented attribute and Strava may add values other than 0 or 1 in the future, I'd choose to be rather forgiving and have the validator implemented by just adding `pre=True` to the existing validator. It'll return our optional literal, or None if encountering another value. I don't see a need to raise an exception in this case.

Reply:
Agree you'll still need the if else construct however to get mypy happy

Reply:
eeks so a few questions - what i have now:

```python
 @validator("athlete_type", pre=True)
    def to_str_representation(
        cls, raw_type: int
    ) -> Optional[Literal["cyclist", "runner"]]:
        """Replaces legacy 'ChoicesAttribute' class.

        Parameters
        ----------
        raw_type : int
            The raw integer representing the athlete type.

        Returns
        -------
        str
            The string representation of the athlete type.
        """

        athlete_type = {0: "cyclist", 1: "runner"}
        if raw_type in [0, 1]:
            return athlete_type.get(raw_type)
        else:
            return None
```

1. should the type for raw_type be `raw_type: Literal[0, 1]` given those are the current supported values?
2. mypy is not happy with any of the options (mine or the one @enadeau presented with the else raising a ValueError. 

you can [see the errors here:](https://github.com/stravalib/stravalib/actions/runs/5837735990/job/15833748620?pr=387#step:11:13)

or here is what it's telling me

> stravalib/model.py:606: error: Incompatible return value type (got "Optional[str]", expected "Optional[Literal['cyclist', 'runner']]")  [return-value]

@enadeau i tried adding that ValueError and i get this

> stravalib/model.py:606: error: Incompatible return value type (got "Optional[str]", expected "Optional[Literal['cyclist', 'runner']]")  [return-value]
> stravalib/model.py:608: error: Statement is unreachable  [unreachable]
> stravalib/model.py:610: error: Name "raiseValueError" is not defined  [name-defined]

i generally understand what is happening here but i don't understand what mypy needs to pass

Reply:
Based on @enadeau 's comment above, I think this should be the validator's logic:

```python
if raw_type == 0:
    return "cyclist"
elif raw_type == 1:
    return "runner"
else:
    return None
```

I see now that the dictionary `.get()` shortcut will not work because the dict itself is not typed, so mypy cannot check if one of the literal values will come out of it.

Let's also keep the `raw_type`'s type as `Any` as this is an undocumented attribute and we don't know what Strava will return in the future. Maybe we could log a warning in the `else` case to indicate we got an unexpected value.

Reply:
ok @enadeau i'm sorry, i didn't understand why we needed the more explicit construct but @jsamoocha now helps me understand that with `.get()` it's hard to type that dictionary. so by being explicit  mypy will "See" that there are two (Literals) options here. for some reason i thought we had to raise that error for mypy. but i get it now. i've also added a LOGGER.warning element as suggested.